ksql-engine/src/main/java/io/confluent/ksql/analyzer/Analyzer.java

```diff
@@ -397,14 +387,95 @@ protected AstNode visitJoin(final Join node, final Void context) {
       return null;
     }
 
-    private boolean validJoin(
-        final SourceName leftName,
-        final SourceName rightName,
+    private void throwOnSelfJoin(final AliasedDataSource left, final AliasedDataSource right) {
+      if (left.getDataSource().getName().equals(right.getDataSource().getName())) {
+        throw new KsqlException(
+            "Can not join '" + left.getDataSource().getName().toString(FormatOptions.noEscape())
+                + "' to '" + right.getDataSource().getName().toString(FormatOptions.noEscape())
+                + "': self joins are not yet supported."
+        );
+      }
+    }
+
+    private void throwOnIncompleteJoinCriteria(
```

Comment:
This checks if the left and right alias are the same? Might want to consolidate it under the self join check. The current name is a bit misleading

Reply:
No, this checks the the two sides of the join criteria reference both sources, e.g. it rejects a join like:

```
... FROM A JOIN B ON B.COL1 = B.COL2;
```

Note: source `A` is not included in the join criteria.